src/iso15118/io/connection_ssl.cpp

```diff
@@ -1,288 +1,370 @@
+        log_and_raise_openssl_error("Failed in SSL_CTX_set_min_proto_version()");
     }
 
-    const auto ssl_setup_result = mbedtls_ssl_setup(&ssl->ssl, &ssl->conf);
-    if (ssl_setup_result != 0) {
-        log_and_raise_mbed_error("Failed to mbedtls_ssl_setup()", ssl_setup_result);
+    if (SSL_CTX_set_cipher_list(ctx, TLS1_2_CIPHERSUITES) == 0) {
+        log_and_raise_openssl_error("Failed in SSL_CTX_set_cipher_list()");
     }
 
-    poll_manager.register_fd(ssl->accepting_net_ctx.fd, [this]() { this->handle_connect(); });
-}
+    if (SSL_CTX_set_ciphersuites(ctx, TLS1_3_CIPHERSUITES) == 0) {
+        log_and_raise_openssl_error("Failed in SSL_CTX_set_ciphersuites()");
+    }
 
-ConnectionSSL::~ConnectionSSL() = default;
+    // TODO(sl): Better cert path handling
+    const std::filesystem::path certificate_chain_file_path = prefix / "client/cso/CPO_CERT_CHAIN.pem";
 
-void ConnectionSSL::set_event_callback(const ConnectionEventCallback& callback) {
-    this->event_callback = callback;
-}
+    if (SSL_CTX_use_certificate_chain_file(ctx, certificate_chain_file_path.c_str()) != 1) {
+        log_and_raise_openssl_error("Failed in SSL_CTX_use_certificate_chain_file()");
+    }
 
-Ipv6EndPoint ConnectionSSL::get_public_endpoint() const {
-    return end_point;
+    // INFO: the password callback uses a non-const argument
+    std::string pass_str = ssl_config.private_key_password;
+    SSL_CTX_set_default_passwd_cb_userdata(ctx, pass_str.data());
+
+    const std::filesystem::path private_key_file_path = prefix / "client/cso/SECC_LEAF.key";
+    if (SSL_CTX_use_PrivateKey_file(ctx, private_key_file_path.c_str(), SSL_FILETYPE_PEM) != 1) {
+        log_and_raise_openssl_error("Failed in SSL_CTX_use_PrivateKey_file()");
+    }
+
+    // TODO(sl): How switch verify mode to none if tls 1.2 is used?
+    SSL_CTX_set_verify(ctx, SSL_VERIFY_NONE, nullptr);
+
+    if (ssl_config.enable_ssl_logging) {
+
+        SSL_CTX_set_keylog_callback(ctx, [](const SSL* ssl, const char* line) {
+            logf("TLS handshake keys: %s\n", line);
+
+            if (key_log_fd != -1) {
+                sockaddr_in6 destination_address = {AF_INET6, htons(key_log_udp_port)};
+                if (inet_pton(AF_INET6, LINK_LOCAL_MULTICAST, &destination_address.sin6_addr) <= 0) {
+                    const auto error_msg = adding_err_msg("Failed to setup server address");
+                    logf(error_msg.c_str());
+                    return;
+                }
+                const auto udp_send_result =
+                    sendto(key_log_fd, line, strlen(line), 0, reinterpret_cast<const sockaddr*>(&destination_address),
+                           sizeof(destination_address));
+
+                if (udp_send_result != strlen(line)) {
+                    const auto error_msg = adding_err_msg("UDP send() failed");
+                    logf(error_msg.c_str());
+                }
+            }
+        });
+    }
```

Comment:
Perhaps this could be a function call for better readability

Reply:
With the `create_udp_socket` refactor, I don't think this needs to be in a separate function.

Reply:
Alright!